pio/isr/main.c

```diff
@@ -0,0 +1,193 @@
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <stdio.h>
+#include "pico/stdlib.h"
+
+#include "simply_isr.pio.h"
+// the sdk doesn't give us functions for figuring out which
+// isr source we're using, so we have to do it manually.
+#include "hardware/regs/pio.h"
+
+/**
+ * These defines represent each state machine.
+ * The value is the bit in the IRQ register that
+ * will be set by each state machine thanks to "irq wait 0 rel"
+ */
+#define PIO_SM_0_IRQ 0b0001
+#define PIO_SM_1_IRQ 0b0010
+#define PIO_SM_2_IRQ 0b0100
+#define PIO_SM_3_IRQ 0b1000
+
+/**
+ * This variable will shadow the IRQ flags set by the PIO state machines.
+ * Typically you do not want to do work in ISRs because the main thread
+ * has more important things to do. Because of that, when we get the ISR
+ * I'm simply going to copy the state machine that fired the ISR into
+ * this variable.
+ *
+ * Variable is volatile so that it doesn't get cached in a CPU register
+ * in the main thread. Without this it's possible that you never see
+ * irq_flags get set even though the ISR is firing all the time.
+ *
+ * Of course, you can really do whatever you want in the ISR, it's up to you.
+ */
+volatile uint32_t irq_flags = 0;
+
+/**
+ * Reads a hardware register
+ *
+ * @note See datasheet and/or pico-sdk/src/rp2040/include/hardware/regs
+ *       for base addresses and offsets
+ * @param[in] bank The base address of the register set usually <MODULE>_BASE
+ * @param[in] offset The register offset, usually <REG_NAME>_OFFSET
+ */
+uint32_t read_register(uint32_t bank, uint32_t offset) {
+    return *((volatile io_rw_32*)(bank + offset));
+}
+
+/**
+ * Writes a hardware register
+ *
+ * @note See datasheet and/or pico-sdk/src/rp2040/include/hardware/regs
+ *       for base addresses and offsets
+ * @param[in] bank The base address of the register set usually <MODULE>_BASE
+ * @param[in] offset The register offset, usually <REG_NAME>_OFFSET
+ * @param[in] value The value to write to the register
+ */
+void write_register(uint32_t bank, uint32_t offset, uint32_t value) {
+    *((volatile io_rw_32*)(bank + offset)) = value;
+}
+
+/**
+ * This function is called when the IRQ is fired by the state machine.
+ * @note See enable_pio_isrs for how to register this function to be called
+ */
+void simply_isr_handler() {
+    // Read the IRQ register to get the IRQ flags from the state machine
+    // This tells me which state machine sent the IRQ
+    irq_flags = read_register(PIO0_BASE, PIO_IRQ_OFFSET);
```

Comment:
Looks like you can also get this from https://github.com/raspberrypi/pico-sdk/blob/master/src/rp2040/hardware_structs/include/hardware/structs/pio.h#L127-L130 which might be "less lowlevel" than defining your own `read_register` and `write_register` functions? :shrug: 

Or there's also https://github.com/raspberrypi/pico-sdk/blob/master/src/rp2_common/hardware_pio/include/hardware/pio.h#L799-L822 which is probably even better?

Reply:
Oh I didn't notice the `pio_interrupt_get`. I think that will work better for this example.

Reply:
Updated to use pio_interrupt_get.
I think personally I would prefer the struct version to read the IRQ, just because it's 1 read vs. 4 separate reads for `pio_interrupt_get`. But as an sdk example I think this is good.